File: leetcode_python/Stack/collecting-chocolates.py
"""

2735. Collecting Chocolates
Medium

You are given a 0-indexed integer array nums of size n representing the cost of collecting different chocolates. The cost of collecting the chocolate at the index i is nums[i]. Each chocolate is of a different type, and initially, the chocolate at the index i is of ith type.

In one operation, you can do the following with an incurred cost of x:

Simultaneously change the chocolate of ith type to ((i + 1) mod n)th type for all chocolates.

Return the minimum cost to collect chocolates of all types, given that you can perform as many operations as you would like.


Example 1:

Input: nums = [20,1,15], x = 5
Output: 13
Explanation: Initially, the chocolate types are [0,1,2]. We will buy the 1st type of chocolate at a cost of 1.
Now, we will perform the operation at a cost of 5, and the types of chocolates will become [1,2,0]. We will buy the 2nd type of chocolate at a cost of 1.
Now, we will again perform the operation at a cost of 5, and the chocolate types will become [2,0,1]. We will buy the 0th type of chocolate at a cost of 1.
Thus, the total cost will become (1 + 5 + 1 + 5 + 1) = 13. We can prove that this is optimal.

Example 2:

Input: nums = [1,2,3], x = 4
Output: 6
Explanation: We will collect all three types of chocolates at their own price without performing any operations. Therefore, the total cost is 1 + 2 + 3 = 6.


Constraints:

1 <= nums.length <= 1000
1 <= nums[i] <= 10^9
1 <= x <= 10^9

"""

# V0
# IDEA : ENUMERATE THE TOTAL NUMBER OF ROTATIONS + RUNNING PREFIX MIN
#
#   the operations are global rotations, and buying is free to happen at any
#   moment, so the ONLY real decision is k = how many rotations we ever do.
#   Fix k: we pay x * k, and for each type t we may buy it at any of the k + 1
#   snapshots, i.e. at price min(nums[t], nums[t-1], ..., nums[t-k]) (indices
#   mod n) — after r rotations the chocolate sitting at position t-r carries
#   type t.
#
#   so, letting best(t, k) = min over that window,
#       answer = min over k in [0, n-1] of ( x * k + sum_t best(t, k) ).
#
#   NOTE : k >= n is never worth it — at k = n-1 the window already covers the
#          whole array, so growing k only adds x per step with no price gain.
#   NOTE : best(t, k) = min(best(t, k-1), nums[(t - k) % n]) — so sweeping k
#          upward per starting index t with ONE running minimum gives O(n^2)
#          time and only O(n) extra space (the full 2D table of the editorial
#          is not needed).
#
# time = O(n^2), space = O(n)   (n <= 1000 -> ~10^6 steps)
class Solution(object):
    def minCost(self, nums, x):
        n = len(nums)
        # cost[k] = total cost if we perform exactly k rotations
        cost = [x * k for k in range(n)]
        for i in range(n):
            cur = nums[i]
            for k in range(n):
                v = nums[(i - k) % n]
                if v < cur:
                    cur = v
                cost[k] += cur
        return min(cost)

File: leetcode_java/src/main/java/LeetCodeJava/Stack/CollectingChocolates.java
package LeetCodeJava.Stack;

// https://leetcode.com/problems/collecting-chocolates/

/**
 *  2735. Collecting Chocolates
 *  Medium
 *
 *  You are given a 0-indexed integer array nums of size n representing the cost
 *  of collecting different chocolates. The cost of collecting the chocolate at
 *  index i is nums[i]. Each chocolate is of a different type, and initially the
 *  chocolate at index i is of ith type.
 *
 *  In one operation, you can do the following with an incurred cost of x:
 *  simultaneously change the chocolate of ith type to ((i + 1) mod n)th type
 *  for all chocolates.
 *
 *  Return the minimum cost to collect chocolates of all types, given that you
 *  can perform as many operations as you would like.
 *
 *  Example 1:
 *    Input: nums = [20,1,15], x = 5
 *    Output: 13
 *    Explanation: buy type 1 for 1, rotate (5), buy type 2 for 1, rotate (5),
 *                 buy type 0 for 1 -> 1 + 5 + 1 + 5 + 1 = 13
 *
 *  Example 2:
 *    Input: nums = [1,2,3], x = 4
 *    Output: 6
 *    Explanation: buy everything at its own price, no operation.
 *
 *  Constraints:
 *    1 <= nums.length <= 1000
 *    1 <= nums[i] <= 10^9
 *    1 <= x <= 10^9
 */
public class CollectingChocolates {

    // V0
    // IDEA: ENUMERATE THE TOTAL NUMBER OF ROTATIONS + RUNNING PREFIX MIN
    //       The operations are global rotations and buying may happen at any
    //       moment, so the only real decision is k = how many rotations we do.
    //       Fix k: we pay x * k, and each type t can be bought at any of the
    //       k + 1 snapshots -> price min(nums[t], nums[t-1], ..., nums[t-k]).
    //       answer = min over k in [0, n-1] of ( x * k + sum_t best(t, k) ).
    //       k >= n never helps (at k = n-1 the window covers the whole array).
    //       best(t, k) = min(best(t, k-1), nums[(t - k) % n]) -> sweep k upward
    //       per starting index with ONE running minimum.
    /**
     * time = O(N^2)
     * space = O(N)
     */
    public long minCost(int[] nums, int x) {
        int n = nums.length;
        long[] cost = new long[n]; // cost[k] = total cost with exactly k rotations
        for (int k = 0; k < n; k++) {
            cost[k] = (long) x * k;
        }
        for (int i = 0; i < n; i++) {
            int cur = nums[i];
            for (int k = 0; k < n; k++) {
                int v = nums[((i - k) % n + n) % n];
                if (v < cur) {
                    cur = v;
                }
                cost[k] += cur;
            }
        }
        long res = cost[0];
        for (int k = 1; k < n; k++) {
            res = Math.min(res, cost[k]);
        }
        return res;
    }
}
